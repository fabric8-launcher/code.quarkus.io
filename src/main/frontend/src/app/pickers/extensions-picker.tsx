import { InputProps, Picker, useAnalytics } from '@launcher/component';
import { FormGroup, TextInput, Tooltip } from "@patternfly/react-core";
import { CheckIcon, ClipboardCheckIcon, ClipboardIcon, SearchIcon, TrashAltIcon } from "@patternfly/react-icons";
import copy from 'copy-to-clipboard';
import React, { useState } from "react";
import './extensions-picker.scss';

export interface ExtensionEntry {
  id: string;
  name: string;
  labels: Set<String>;
  description?: string;
  shortName?: string;
  category: string;
}

export interface ExtensionsPickerValue {
  extensions?: string[];
}

interface ExtensionsPickerProps extends InputProps<ExtensionsPickerValue> {
  entries: ExtensionEntry[];
  placeholder: string;
  filterFunction?(d: ExtensionEntry): boolean;
}

interface ExtensionProps extends ExtensionEntry {
  selected: boolean;
  detailed?: boolean;
  onClick(id: string): void;
}

function CopyToClipboard(props: { content: string }) {
  const [active, setActive] = useState(false);
  const [copied, setCopied] = useState(false);

  const copyToClipboard = () => {
    copy(props.content);
    setCopied(true);
  }
  return (
    <div
      onMouseEnter={() => setActive(true)}
      onMouseLeave={() => setActive(false)}
      onClick={copyToClipboard}

    >
      {active || copied ? <ClipboardCheckIcon /> : <ClipboardIcon />}
    </div>
  )

}

function Extension(props: ExtensionProps) {
  const [active, setActive] = useState(false);
  const onClick = () => {
    props.onClick(props.id);
    setActive(false);
  };

  return (
    <div
      className={`${active ? 'active' : ''} ${props.selected ? 'selected' : ''} extension-item`}
      onMouseEnter={() => setActive(true)}
      onMouseLeave={() => setActive(false)}
      aria-label={`Switch ${props.id} extension`}
    >
      {props.detailed && (
        <div className="extension-selector" onClick={onClick}>
          {((props.selected && !active) || (!props.selected && active)) && <CheckIcon />}
          {(props.selected && active) && <TrashAltIcon />}
        </div>
      )}
      <Tooltip position="top" content={props.name} exitDelay={0}>
        <div className="extension-name" onClick={onClick}>{props.name}</div>
      </Tooltip>
      {!props.detailed && (
        <div className="extension-remove" onClick={onClick}>
          {active && props.selected && <TrashAltIcon />}
        </div>
      )}
      {props.detailed && (
        <div className="extension-details">
          <Tooltip position="top" content={props.description} exitDelay={0}>
            <div className="extension-description" onClick={onClick}>{props.description}</div>
          </Tooltip>
          <Tooltip position="left" content={`Copy '${props.id}' to clipboard`} exitDelay={0}>
            <div className="extension-gav"><CopyToClipboard content={props.id} /></div>
          </Tooltip>
        </div>
      )}
    </div>
  )
}

export const filterFunction = (filter: string, shortNames: Set<String>) => (d: ExtensionEntry) => {
  const filterLowerCase = filter.toLowerCase();
  if (!filterLowerCase) {
    return true;
  }
  const shortName = d.shortName ? d.shortName.toLowerCase() : '';
  if (filterLowerCase === shortName) {
    return true;
  }
  if (shortNames.has(filterLowerCase)) {
    return false;
  }
  return d.name.toLowerCase().includes(filterLowerCase)
    || d.labels.has(filterLowerCase)
    || (d.category && d.category.toLowerCase().startsWith(filterLowerCase))
    || shortName.startsWith(filterLowerCase);
}

export function getShortNames(entries: ExtensionEntry[]) {
  const shortNames = entries
    .map(e => e.shortName && e.shortName.toLowerCase())
    .filter(e => !!e) as string[];
  return new Set(shortNames);
}

export const sortFunction = (filter: string) => (a: ExtensionEntry, b: ExtensionEntry) => {
  const filterLowerCase = filter.toLowerCase();
  if (!filterLowerCase) {
    if (a.category === b.category) {
      return a.name > b.name ? -1 : 1;
    }
    return a.category > b.category ? -1 : 1;
  }
  const startWithAShortName = !!a.shortName && a.shortName.toLowerCase().startsWith(filterLowerCase);
  const startWithBShortName = !!b.shortName && b.shortName.toLowerCase().startsWith(filterLowerCase);
  if (startWithAShortName !== startWithBShortName) {
    return startWithAShortName ? -1 : 1;
  }
  if (a.labels.has(filterLowerCase) !== b.labels.has(filterLowerCase)) {
    return a.labels.has(filterLowerCase) ? -1 : 1;
  }
  if (a.name.toLowerCase().startsWith(filterLowerCase) !== b.name.toLowerCase().startsWith(filterLowerCase)) {
    return a.name.toLowerCase().startsWith(filterLowerCase) ? -1 : 1;
  }
  return a.name < b.name ? -1 : 1;
}

export const ExtensionsPicker: Picker<ExtensionsPickerProps, ExtensionsPickerValue> = {
  checkCompletion: (value: ExtensionsPickerValue) => !!value.extensions && value.extensions.length > 0,
  Element: (props: ExtensionsPickerProps) => {
    const [filter, setFilter] = useState('');
    const analytics = useAnalytics();
    const extensions = props.value.extensions || [];
    const entrySet = new Set(extensions);
    const entriesById: Map<String, ExtensionEntry> = new Map(props.entries.map(item => [item.id, item]));

    const add = (id: string) => {
      entrySet.add(id);
      props.onChange({ extensions: Array.from(entrySet) });
      analytics.event('Picker', 'Add-Extension', id);
    };

    const remove = (id: string) => {
      entrySet.delete(id);
      props.onChange({ extensions: Array.from(entrySet) });
      analytics.event('Picker', 'Remove-Extension', id)
    };

    const shortNames = getShortNames(props.entries);

    const result = props.entries.filter(filterFunction(filter, shortNames));
    const categories = new Set(props.entries.map(i => i.category));
    let currentCat: string | undefined = undefined;
    return (
      <div className="extensions-picker">
        <div className="control-container">
          <Tooltip position="right" content={`${Array.from(categories).join(', ')}`}>
            <FormGroup
              fieldId="search-extensions-input"
            >
              <SearchIcon />
              <TextInput
                aria-label="Search extensions"
                placeholder={props.placeholder}
                className="search-extensions-input"
                value={filter}
                onChange={value => setFilter(value)}
              />
            </FormGroup>
          </Tooltip>
          <div className={`selected-extensions`}>
            <h4>Selected Extensions</h4>
            <div className={`extension-list`}>
              {
                extensions.map((ex, i) => (
                  <Extension
                    selected={entrySet.has(ex)}
                    {...entriesById.get(ex)!}
                    key={i}
                    onClick={entrySet.has(ex) ? remove : add}
                  />
                ))
              }
            </div>
          </div>
        </div>
        <div className="list-container">
          {result.sort(sortFunction(filter)).map((ex, i) => {
            const ext = (
              <Extension
                selected={entrySet.has(ex.id)}
                {...ex}
                key={i}
                onClick={entrySet.has(ex.id) ? remove : add}
                detailed
              />
            );
            if (!filter && (!currentCat || currentCat !== ex.category)) {
              currentCat = ex.category;
              return (
                <div style={{display: 'contents'}}>
                  <div className="extension-category">
                    {currentCat}
                  </div>
                  {ext}
                </div>
              )
            }
            return ext;
          })}
        </div>
      </div>

    );
  }
}