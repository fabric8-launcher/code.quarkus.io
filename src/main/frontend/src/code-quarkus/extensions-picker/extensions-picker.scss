@import "../config.scss";

.extensions-picker {
  color: $extensionsPickerTextColor;

  .main-container .extension-list-wrapper {
    .extension-category {
      border-bottom: 1px solid $extensionsPickerCategoryUnderlineColor;
      font-weight: bold;
      color: $extensionsPickerCategoryTextColor;
      font-size: 1.1rem;
      margin: 10px 0 5px 0;
      height: 30px;
    }

    .extension-category:first-child {
      margin: 0 0 5px 0;
    }

    .extension-row {
      color: $extensionsPickerTextColor;

      &.keyboard-actived {
        color: $extensionsPickerKeyboardActivedTextColor;
      }

      .extension-selector {
        flex-basis: 30px;
        order: 1;

        svg {
          vertical-align: middle;
        }
      }

      .extension-description {
        flex: 1;
        order: 4;
        line-height: 1.2rem;
        flex-basis: 100%;
      }

      .extension-summary {
        flex-basis: 350px;
        order: 2;
        flex-grow: 1;
        flex-basis: calc(100% - 62px);

        @media screen and (max-width: $mobileWidth) {
          flex-basis: calc(100% - 32px);
        }
      }

      .extension-more {
        order: 5;
        order: 3;
        flex-basis: 32px;
      }

      &.readonly {
        cursor: initial;
      }
    }
  }

  .control-container {
    position: sticky;
    top: 200px;
    display: flex;
    background-color: $mainContainerBg;
    z-index: 99;
    padding-top: 20px;

    .search-bar {
      margin: 0 auto;
      width: 100%
    }

    @media screen and (max-width: $mobileWidth) {
      display: none;
    }

    .form-group {
      svg {
        margin: 0 10px;
        width: 20px;
        vertical-align: middle;
      }
    }

    input.form-control {
      border: 1px solid $extensionsPickerSearchBorderColor;
      padding-left: 40px;
      margin-left: -40px;
      font-size: 14px;
      height: 40px;
      line-height: 40px;
      width: 100%;
      background: transparent;
      color: $extensionsPickerSearchTextColor;
      font-family: $inputFont;
      padding-bottom: 4px;
    }

    input.form-control:focus {
      outline: none;
      border: 1px solid $extensionsPickerSearchBorderColorOnFocus;
    }

    h4 {
      border-bottom: 1px solid $mainContainerTextColor;
      font-weight: bold;
      margin-top: 40px;
      margin-bottom: 20px;
    }

    .selected-extensions {
      flex-grow: 1;
      overflow: hidden;
    }

    .extension-summary {
      flex-grow: 1;
    }

    .extension-list-wrapper {
      height: calc(100% - 25px);
      overflow-y: auto;
    }
  }

  .main-container {
    margin: 0 auto;
    padding: 20px 20px 200px;

    flex-grow: 1;
    overflow: hidden;

    @media screen and (max-width: $mobileWidth) {
      margin: 5px;
      padding: 0px 0px 200px;
    }

    .extension-search-clear {
      border-bottom: 1px solid $extensionsPickerCategoryUnderlineColor;
      font-weight: bold;
      font-size: 1.1rem;
      margin: 0 0 5px 0;
      height: 30px;

      a.btn {
        cursor: pointer;
        color: $extensionsPickerSearchClearColor;
        padding: 0 3px;
      }
    }
  }
}
