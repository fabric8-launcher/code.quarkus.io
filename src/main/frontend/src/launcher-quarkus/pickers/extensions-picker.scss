.extensions-picker {
  display: flex;
  color: white;

  .control-container {
    flex: 0 0 350px;

    .pf-c-form__group {
      svg {
        margin: 0 10px;
      }
    }

    input.pf-c-form-control {
      border: 1px solid #9aff2d;
      padding-left: 40px;
      margin-left: -40px;
      font-size: 14px;
      width: 100%;
      background: transparent;
      color: white;
      font-family: "PT Mono", Regular;
      padding-bottom: 4px;
    }

    input.pf-c-form-control:focus {
      outline: none;
      border: 1px solid #4695eb;
    }

    h4 {
      border-bottom: 1px solid #ff004a;
      font-weight: bold;
      margin-top: 40px;
      margin-bottom: 20px;
    }
    .extension-list {
      display: grid;
      grid-template-columns: auto 30px;
      .extension-item {
        cursor: pointer;
        display: contents;
      }
    }
  }

  .extension-item {
    cursor: pointer;

    .extension-selector {
      color: #fff;
    }

    &.selected {
      .extension-selector {
        color: #71aeef;
      }
    }

    &.active.selected {
      .extension-selector {
        color: #fff;
      }
    }

    .extension-remove {
      color: #ff004a;
    }

    & div:focus {
      outline: none;
    }

    .extension-selector {
      max-width: 10px;

      @media screen and (max-width: 900px) {
        display: none;
      }
    }

    .extension-name {
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
      font-size: 0.9rem;
      font-weight: bold;
      height: 24px;
    }

    .extension-details {
      display: contents;
    }

    .extension-description {
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
      font-size: 0.8rem;
    }

    .extension-gav {
      max-width: 150px;
      direction: rtl;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
      text-align: right;
      font-size: 0.8rem;

      @media screen and (max-width: 900px) {
        display: none;
      }
    }
  }
}

.result-container {
  margin-left: 30px;
  flex-grow: 1;

  @media screen and (max-width: 900px) {
    margin: 5px;
  }

  .extension-search-clear {
    border-bottom: 1px solid #4695eb;
    font-weight: bold;
    font-size: 1.1rem;
    margin: 0 0 5px 0;
    height: 30px;
    .pf-c-button.pf-m-link {
      color: #ff004a;
      padding: 0 3px;
    }
  }
}

.list-container {
  display: grid;
  grid-template-columns: 30px 300px auto 30px;

  @media screen and (max-width: 900px) {
    grid-template-columns: 150px auto;
  }

  .extension-category {
    grid-column: 1 / span 4;
    border-bottom: 1px solid #4695eb;
    font-weight: bold;
    font-size: 1.1rem;
    margin: 10px 0 5px 0;
    height: 30px;

    @media screen and (max-width: 900px) {
      grid-column: 1 / span 2;
    }
  }

  .extension-category:first-child {
    margin: 0 0 5px 0;
  }

  .extension-item {
    cursor: pointer;
    color: white;
    display: contents;

    & > div {
      margin: 0 10px;
    }
  }
}

.quarkus-blurb {
  margin-bottom: 20px;
}