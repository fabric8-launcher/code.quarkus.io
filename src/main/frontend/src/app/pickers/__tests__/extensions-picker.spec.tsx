import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { ExtensionsPicker, filterFunction, ExtensionEntry, getShortNames, sortFunction } from '../extensions-picker';

afterEach(() => {
  cleanup();
});

const entries: ExtensionEntry[] = [
  {
    "id": "io.quarkus:quarkus-arc",
    "name": "ArC",
    "labels": new Set([
      "arc",
      "cdi",
      "dependency-injection",
      "di",
      "label"
    ]),
    "description": "Build time CDI dependency injection",
    "shortName": "CDI",
    "category": "Core"
  },
  {
    "id": "io.quarkus:quarkus-camel-netty4-http",
    "name": "Camel Netty4 test HTTP",
    "labels": new Set([
      "camel-netty4-http",
      "camel"
    ]),
    "description": "Camel support for Netty",
    "category": "Integration"
  },
  {
    "id": "some-id",
    "name": "A CDI in name test",
    "labels": new Set([
      "lambda",
      "amazon-lambda",
      "aws-lambda",
      "amazon",
      "aws",
      "label"
    ]),
    "shortName": "a shortname",
    "description": "Some description",
    "category": "Cloud"
  },
];

const shortNames = getShortNames(entries);

describe('<ExtensionsPicker />', () => {

  it('renders the ExtensionsPicker correctly', () => {
    const comp = render(<ExtensionsPicker.Element placeholder="" entries={entries} value={{}} onChange={() => { }} />);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show results for valid search', async () => {
    const comp = render(<ExtensionsPicker.Element placeholder="" entries={entries} value={{}} onChange={() => { }} />);

    const searchField = comp.getByLabelText('Search extensions');
    fireEvent.change(searchField, { target: { value: 'CDI' } });
    const result = await comp.findAllByText(entries[0].description!);
    expect((result as HTMLElement[]).length).toBe(1);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('select values and save', async () => {
    const handleChange = jest.fn();
    const comp = render(<ExtensionsPicker.Element placeholder="" entries={entries} value={{}} onChange={handleChange} />);

    const searchField = comp.getByLabelText('Search extensions');
    fireEvent.change(searchField, { target: { value: 'netty' } });
    const item = await comp.findAllByText(entries[1].description!);
    fireEvent.click(item[0]);
    expect(comp.asFragment()).toMatchSnapshot();
  });
});

describe('getShortNames', () => {
  it('should work correctly', () => {
    expect(shortNames).toEqual(new Set(['cdi', 'a shortname']));
  });
})

describe('filterFunction', () => {
  it('when not using a filter, all extensions should be returned', () => {
    expect(entries.filter(filterFunction('', shortNames)))
      .toEqual(entries);
  });

  it('when using exact shortname, only the corresponding extension should be returned', () => {
    expect(entries.filter(filterFunction('cdi', shortNames)))
      .toEqual([entries[0]]);
  });

  it('when using start of shortname, the result should include the other corresponding extensions too', () => {
    expect(entries.filter(filterFunction('cd', shortNames)))
      .toEqual([entries[0], entries[2]]);
  });

  it('when using part of name, it should return the corresponding extensions', () => {
    expect(entries.filter(filterFunction('test', shortNames)))
      .toEqual([entries[1], entries[2]]);
  });

  it('when using label, it should return the corresponding extensions', () => {
    expect(entries.filter(filterFunction('label', shortNames)))
      .toEqual([entries[0], entries[2]]);
  });

  it('when using part of category (not start), it should not return it', () => {
    expect(entries.filter(filterFunction('oud', shortNames)))
      .toEqual([]);
  });

  it('when using start of category, it should return it', () => {
    expect(entries.filter(filterFunction('clou', shortNames)))
      .toEqual([entries[2]]);
  });

  it('when using start of category, it should return it', () => {
    expect(entries.filter(filterFunction('clou', shortNames)))
      .toEqual([entries[2]]);
  });
});

describe('sortFunction', () => {
  it('when using start of shortname of an extension, it should be first', () => {
    expect(sortFunction('cdi')(entries[0], entries[2]))
      .toEqual(-1);
    expect(sortFunction('cdi')(entries[2], entries[0]))
      .toEqual(1);
  });

  it('when using one of the label of an extension, it should be first', () => {
    expect(sortFunction('amazon')(entries[2], entries[0]))
      .toEqual(-1);
    expect(sortFunction('amazon')(entries[0], entries[2]))
      .toEqual(1);
  });

  it('when using start of the name of an extension, it should be first', () => {
    expect(sortFunction('a')(entries[0], entries[1]))
      .toEqual(-1);
    expect(sortFunction('a')(entries[1], entries[0]))
      .toEqual(1);
  });

  it('when no match, it should compare the names', () => {
    expect(sortFunction('nomatch')(entries[0], entries[1]))
      .toEqual(-1);
    expect(sortFunction('nomatch')(entries[1], entries[0]))
      .toEqual(1);
  });

});