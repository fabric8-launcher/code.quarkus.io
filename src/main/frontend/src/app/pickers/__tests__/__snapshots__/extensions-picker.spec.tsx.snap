// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExtensionsPicker /> renders the ExtensionsPicker correctly 1`] = `
<DocumentFragment>
  <div
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="pf-c-form__group"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            transform=""
          />
        </svg>
        <input
          aria-invalid="false"
          aria-label="Search extensions"
          class="pf-c-form-control search-extensions-input"
          placeholder=""
          type="text"
          value=""
        />
      </div>
      <div
        class="selected-extensions"
      >
        <h4>
          Selected Extensions
        </h4>
        <div
          class="extension-list"
        />
      </div>
    </div>
    <div
      class="list-container"
    >
      <div
        style="display: contents;"
      >
        <div
          class="extension-category"
        >
          Integration
        </div>
        <div
          aria-label="Switch io.quarkus:quarkus-camel-netty4-http extension"
          class="  extension-item"
        >
          <div
            class="extension-selector"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                transform=""
              />
            </svg>
          </div>
          <div
            class="extension-name"
            tabindex="0"
          >
            Camel Netty4 test HTTP
          </div>
          <div
            class="extension-details"
          >
            <div
              class="extension-description"
              tabindex="0"
            >
              Camel support for Netty
            </div>
            <div
              class="extension-gav"
              tabindex="0"
            >
              <div>
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                    transform=""
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="display: contents;"
      >
        <div
          class="extension-category"
        >
          Core
        </div>
        <div
          aria-label="Switch io.quarkus:quarkus-arc extension"
          class="  extension-item"
        >
          <div
            class="extension-selector"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                transform=""
              />
            </svg>
          </div>
          <div
            class="extension-name"
            tabindex="0"
          >
            ArC
          </div>
          <div
            class="extension-details"
          >
            <div
              class="extension-description"
              tabindex="0"
            >
              Build time CDI dependency injection
            </div>
            <div
              class="extension-gav"
              tabindex="0"
            >
              <div>
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                    transform=""
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="display: contents;"
      >
        <div
          class="extension-category"
        >
          Cloud
        </div>
        <div
          aria-label="Switch some-id extension"
          class="  extension-item"
        >
          <div
            class="extension-selector"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                transform=""
              />
            </svg>
          </div>
          <div
            class="extension-name"
            tabindex="0"
          >
            A CDI in name test
          </div>
          <div
            class="extension-details"
          >
            <div
              class="extension-description"
              tabindex="0"
            >
              Some description
            </div>
            <div
              class="extension-gav"
              tabindex="0"
            >
              <div>
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                    transform=""
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ExtensionsPicker /> select values and save 1`] = `
<DocumentFragment>
  <div
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="pf-c-form__group"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            transform=""
          />
        </svg>
        <input
          aria-invalid="false"
          aria-label="Search extensions"
          class="pf-c-form-control search-extensions-input"
          placeholder=""
          type="text"
          value="netty"
        />
      </div>
      <div
        class="selected-extensions"
      >
        <h4>
          Selected Extensions
        </h4>
        <div
          class="extension-list"
        />
      </div>
    </div>
    <div
      class="list-container"
    >
      <div
        class="extension-search-clear"
      >
        Search results (
        <button
          class="pf-c-button pf-m-link"
          type="button"
        >
          Clear search
        </button>
        )
      </div>
      <div
        aria-label="Switch io.quarkus:quarkus-camel-netty4-http extension"
        class="  extension-item"
      >
        <div
          class="extension-selector"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              transform=""
            />
          </svg>
        </div>
        <div
          class="extension-name"
          tabindex="0"
        >
          Camel Netty4 test HTTP
        </div>
        <div
          class="extension-details"
        >
          <div
            class="extension-description"
            tabindex="0"
          >
            Camel support for Netty
          </div>
          <div
            class="extension-gav"
            tabindex="0"
          >
            <div>
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 384 512"
                width="1em"
              >
                <path
                  d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                  transform=""
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ExtensionsPicker /> show results for valid search 1`] = `
<DocumentFragment>
  <div
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="pf-c-form__group"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            transform=""
          />
        </svg>
        <input
          aria-invalid="false"
          aria-label="Search extensions"
          class="pf-c-form-control search-extensions-input"
          placeholder=""
          type="text"
          value="CDI"
        />
      </div>
      <div
        class="selected-extensions"
      >
        <h4>
          Selected Extensions
        </h4>
        <div
          class="extension-list"
        />
      </div>
    </div>
    <div
      class="list-container"
    >
      <div
        class="extension-search-clear"
      >
        Search results (
        <button
          class="pf-c-button pf-m-link"
          type="button"
        >
          Clear search
        </button>
        )
      </div>
      <div
        aria-label="Switch io.quarkus:quarkus-arc extension"
        class="  extension-item"
      >
        <div
          class="extension-selector"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              transform=""
            />
          </svg>
        </div>
        <div
          class="extension-name"
          tabindex="0"
        >
          ArC
        </div>
        <div
          class="extension-details"
        >
          <div
            class="extension-description"
            tabindex="0"
          >
            Build time CDI dependency injection
          </div>
          <div
            class="extension-gav"
            tabindex="0"
          >
            <div>
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 384 512"
                width="1em"
              >
                <path
                  d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
                  transform=""
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
